python/ray/serve/_private/metrics_utils.py

```diff
@@ -1,138 +1,94 @@
-                argument, and the timestamp of the call is passed as the second
-                argument.
-        """
-        self.task_func: Callable[[], Dict[str]] = task_func
-        self.interval_s: float = interval_s
-        self.callback_func: Callable[[Any, float]] = callback_func
-        self.last_ref: Optional[ray.ObjectRef] = None
-        self.last_call_succeeded_time: Optional[float] = time.time()
+@dataclass
+class _MetricsTask:
+    task_func: Callable
+    interval_s: float
 
 
 class MetricsPusher:
-    """
-    Metrics pusher is a background thread that run the registered tasks in a loop.
-    """
+    """Periodically runs registered asyncio tasks."""
 
-    # If no tasks are registered, sleep for 1s between iterations (until one is
-    # registered).
-    NO_TASKS_REGISTERED_INTERVAL_S = 1
+    def __init__(
+        self,
+        event_loop: asyncio.AbstractEventLoop,
+        async_sleep: Optional[Callable] = None,
+    ):
+        self._event_loop = event_loop
+        self._async_sleep = async_sleep or asyncio.sleep
+        self._tasks: Dict[str, _MetricsTask] = dict()
+        self._async_tasks: Dict[str, asyncio.Task] = dict()
 
-    def __init__(self):
-        self.tasks: Dict[str, _MetricTask] = dict()
-        self.pusher_thread: Union[threading.Thread, None] = None
-        self.stop_event = threading.Event()
+        self._stop_event = asyncio.Event(loop=event_loop)
+        self._wait_on_stop_event = self._event_loop.create_task(self._stop_event.wait())
+
+    def start(self):
+        self._stop_event.clear()
+
+    async def metrics_task(self, name: str):
+        """Periodically runs `task_func` every `interval_s` until `stop_event` is set.
+
+        If `task_func` raises an error, an exception will be logged.
+        """
+
+        while True:
+            if self._stop_event.is_set():
+                return
+
+            try:
+                self._tasks[name].task_func()
+            except Exception as e:
+                logger.exception(f"Failed to run metrics task: {e}")
+
+            sleep_coroutine = self._async_sleep(self._tasks[name].interval_s)
+            await asyncio.wait(
+                [self._wait_on_stop_event, asyncio.create_task(sleep_coroutine)],
+                return_when=asyncio.FIRST_COMPLETED,
+            )
 
     def register_or_update_task(
         self,
         name: str,
         task_func: Callable,
         interval_s: int,
-        process_func: Optional[Callable] = None,
     ) -> None:
         """Register a task under the provided name, or update it.
 
         This method is idempotent - if a task is already registered with
         the specified name, it will update it with the most recent info.
         """
 
-        self.tasks[name] = _MetricTask(task_func, interval_s, process_func)
+        self._tasks[name] = _MetricsTask(task_func, interval_s)
+        if name not in self._async_tasks:
```

Comment:
Seems like `self._async_tasks` is only used here to check if the task are already assigned? If so, we just need it to be a set of string (but let me know if we need to keep the references as well)

Reply:
Maybe we do need to wait/ cancel for all those with references. So this becomes noop.